astroplan/constraints.py

```diff
@@ -0,0 +1,303 @@
+        altaz = cached_altaz['altaz']
+        if self.min is None:
+            mask = altaz.secz < self.max
+        elif self.max is None:
+            mask = self.min < altaz.secz
+        elif self.min is not None and self.max is not None:
+            mask = (self.min < altaz.secz) & (altaz.secz < self.max)
+        return mask
+
+class AtNight(Constraint):
+    """
+    Constrain the Sun to be below ``horizon``.
+    """
+    @u.quantity_input(horizon=u.deg)
+    def __init__(self, max_solar_altitude=0*u.deg, force_pressure_zero=True):
+        """
+        Parameters
+        ----------
+        max_solar_altitude : `~astropy.units.Quantity`
+            Define "night" as when the sun is below ``max_solar_altitude``.
+            Default is zero degrees altitude.
+        """
+        self.max_solar_altitude = max_solar_altitude
+        self.force_pressure_zero = force_pressure_zero
+
+    @classmethod
+    def twilight_civil(cls, **kwargs):
+        return cls(max_solar_altitude=-6*u.deg, **kwargs)
+
+    @classmethod
+    def twilight_nautical(cls, **kwargs):
+        return cls(max_solar_altitude=-12*u.deg, **kwargs)
+
+    @classmethod
+    def twilight_astronomical(cls, **kwargs):
+        return cls(max_solar_altitude=-18*u.deg, **kwargs)
+
+    def _get_solar_altitudes(self, time_range, observer, targets,
+                             time_resolution=DEFAULT_TIME_RESOLUTION):
+        if not hasattr(observer, '_altaz_cache'):
+            observer._altaz_cache = {}
+
+        times = Time(np.arange(time_range[0].jd, time_range[1].jd,
+                               time_resolution.to(u.day).value), format='jd')
+
+        aakey = (tuple(times.jd), 'sun')
+
+        if aakey not in observer._altaz_cache:
+            if self.force_pressure_zero:
+                observer_old_pressure = observer.pressure
+                observer.pressure = 0
+
+            # Broadcast the solar altitudes for the number of targets:
+            altaz = observer.altaz(times, get_sun(times))
+            altitude = altaz.alt
+            altitude.resize(1, len(altitude))
+            altitude = altitude + np.zeros((len(targets), 1))
+
+            observer._altaz_cache[aakey] = dict(times=times,
+                                                altitude=altitude)
+            if self.force_pressure_zero:
+                observer.pressure = observer_old_pressure
+
+        return observer._altaz_cache[aakey]
+
+    def _compute_constraint(self, time_range, observer, targets):
+        sun_altaz = self._get_solar_altitudes(time_range, observer, targets)
+        solar_altitude = sun_altaz['altitude']
+        mask = solar_altitude < self.max_solar_altitude
+        return mask
+
+def is_always_observable(constraints, time_range, targets, observer):
+    """
+    Are the ``targets`` always observable throughout ``time_range`` given
+    constraints in ``constraints_list`` for ``observer``?
```

Comment:
Looks like "constraints_list" should be changed to "constraints" to match the code?